demos/digest/EVP_MD_xof.c

```diff
@@ -0,0 +1,145 @@
+ *
+ * To generate different output lengths, you can pass a single integer argument
+ * on the command line, which is the output size in bytes. By default, a 20-byte
+ * output is generated and (for this length only) a known answer test is
+ * performed.
+ */
+
+/* Our input to the XOF hash function. */
+const char message[] = "This is a test message.";
+
+/* Expected output when an output length of 20 bytes is used. */
+static const char known_answer[] = {
+  0x52, 0x97, 0x93, 0x78, 0x27, 0x58, 0x7d, 0x62,
+  0x8b, 0x00, 0x25, 0xb5, 0xec, 0x39, 0x5e, 0x2d,
+  0x7f, 0x3e, 0xd4, 0x19
+};
+
+/*
+ * A property query used for selecting the SHAKE256 implementation.
+ */
+static const char *propq = NULL;
+
+int main(int argc, char **argv)
+{
+    int rv = 1;
+    OSSL_LIB_CTX *libctx = NULL;
+    EVP_MD *md = NULL;
+    EVP_MD_CTX *ctx = NULL;
+    OSSL_PARAM params[2], *p = params;
+    unsigned int digest_len = 20;
+    int digest_len_i;
+    unsigned char *digest = NULL;
+
+    /* Allow digest length to be changed for demonstration purposes. */
+    if (argc > 1) {
+        digest_len_i = atoi(argv[1]);
+        if (digest_len_i <= 0) {
+            fprintf(stderr, "Specify a non-negative digest length\n");
+            goto end;
+        }
+
+        digest_len = (unsigned int)digest_len_i;
+    }
+
+    /*
+     * Retrieve desired algorithm. This must be a hash algorithm which supports
+     * XOF.
+     */
+    md = EVP_MD_fetch(libctx, "SHAKE256", propq);
+    if (md == NULL) {
+        fprintf(stderr, "Failed to retrieve SHAKE256 algorithm\n");
+        goto end;
+    }
+
+    /* Create context. */
+    ctx = EVP_MD_CTX_new();
+    if (ctx == NULL) {
+        fprintf(stderr, "Failed to create digest context\n");
+        goto end;
+    }
+
+    /* Initialize digest context. */
+    if (EVP_DigestInit(ctx, md) == 0) {
+        fprintf(stderr, "Failed to initialize digest\n");
+        goto end;
+    }
+
+    /*
+     * Set the length of the output we desire.
+     * Note that the output length is specified in bytes.
+     */
+    *p++ = OSSL_PARAM_construct_uint(OSSL_DIGEST_PARAM_XOFLEN, &digest_len);
+    *p++ = OSSL_PARAM_construct_end();
+
+    if (EVP_MD_CTX_set_params(ctx, params) == 0) {
+        fprintf(stderr, "Failed to set XOF length\n");
+        goto end;
+    }
+
+    /*
+     * Feed our message into the digest function.
+     * This may be called multiple times.
+     */
+    if (EVP_DigestUpdate(ctx, message, sizeof(message)) == 0) {
+        fprintf(stderr, "Failed to hash input message\n");
+        goto end;
+    }
+
+    /* Allocate enough memory for our digest length. */
+    digest = OPENSSL_malloc(digest_len);
+    if (digest == NULL) {
+        fprintf(stderr, "Failed to allocate memory for digest\n");
+        goto end;
+    }
+
+    /* Get computed digest. */
+    if (EVP_DigestFinal(ctx, digest, &digest_len) == 0) {
+        fprintf(stderr, "Failed to finalize hash\n");
+        goto end;
+    }
```

Comment:
I think setting the XOFLEN and using DigestFinal is the less common way how to generate the XOF output. There is also the EVP_DigestFinalXOF() which allows setting the xof length directly as a parameter.

In future (for implementations that allow this) the EVP_DigestFinalXOF() should even allow multiple calls to produce further XOF bytes.

Reply:
Updated, TIL.

Out of curiosity, what if we ever have to support a XOF which needs the output length to be specified up front (for example, if the length is placed in an internal tweak)?

Reply:
It is not a true XOF then. That's just a variable length hash function. But yeah you'd use the way you wrote the demo originally.